3.evaluate_model/scripts/nbconverted/get_LOIO_probabilities.py

```diff
@@ -0,0 +1,168 @@
+for model_path in sorted(models_dir.iterdir()):
+    # only perform LOIO with hyper params from final models
+    if "shuffled" in model_path.name:
+        continue
+
+    # load the model
+    model = load(model_path)
+    # determine feature type from model file name
+    feature_type = model_path.name.split("__")[1].replace(".joblib", "")
+
+    print(
+        f"Performing LOIO for feature type {feature_type} with parameters C: {model.C}, l1_ratio: {model.l1_ratio}"
+    )
+
+    # iterate through image paths
+    for image_path in labeled_data["Metadata_DNA"].unique():
+        # get training and testing cells from image path
+        # every cell from the image path is for testing, the rest are for training
+        train_cells = labeled_data.loc[labeled_data["Metadata_DNA"] != image_path]
+        test_cells = labeled_data.loc[labeled_data["Metadata_DNA"] == image_path]
+
+        # get X, y from training and testing cells
+        X_train, y_train = get_X_y_data(train_cells, feature_type)
+        X_test, y_test = get_X_y_data(test_cells, feature_type)
+
+        # capture convergence warning from sklearn
+        # this warning does not affect the model but takes up lots of space in the output
+        with parallel_backend("multiprocessing"):
+            with warnings.catch_warnings():
+                warnings.filterwarnings(
+                    "ignore", category=ConvergenceWarning, module="sklearn"
+                )
+
+                # fit a logisitc regression model on the training X, y
+                LOIO_model = LogisticRegression(
+                    penalty="elasticnet",
+                    solver="saga",
+                    max_iter=100,
+                    n_jobs=-1,
+                    random_state=0,
+                    C=model.C,
+                    l1_ratio=model.l1_ratio,
+                ).fit(X_train, y_train)
+
+        # create metadata dataframe for test cells with model parameters
+        metadata_dataframe = pd.concat(
+            [
+                test_cells["Cell_UUID"],
+                test_cells["Metadata_DNA"],
+                test_cells["Mitocheck_Phenotypic_Class"],
+            ],
+            axis=1,
+        ).reset_index(drop=True)
+        metadata_dataframe["Model_Feature_Type"] = feature_type
+        metadata_dataframe["Model_C"] = model.C
+        metadata_dataframe["Model_l1_ratio"] = model.l1_ratio
+
+        # predict probabilities for test cells and make these probabilities into a dataframe
+        probas = LOIO_model.predict_proba(X_test)
+        probas_dataframe = pd.DataFrame(probas, columns=model.classes_)
+
+        # combine metadata and probabilities dataframes for test cells to create wide data
+        test_cells_wide_data = pd.concat([metadata_dataframe, probas_dataframe], axis=1)
+
+        # add tidy long data to compiled data
+        compiled_LOIO_wide_data.append(test_cells_wide_data)
+
+        # DELETE THE REST OF THE LINES IN THIS NOTEBOOK BEFORE FINAL RUN
```

Comment:
Consider removing this comment or providing more details here to denote how this may matter for the audience.

Reply:
This will be removed in the next commit. The notebook takes a while to run (~9hrs) so I wanted to get approval before running everything.